### ISSUE TITLE
Tensor Based Decomposition

### ISSUE THREAD

**Issue by CSSFrancis** [hyperspy/hyperspy#2392]
Recently @Cy-Zhang has been working on a Higher Order-SVD (HOSVD) method for noise reduction in 4-D datasets. (@Cy-Zhang correct me if anything I say is wrong)

This takes into account the inherent periodicity of the lattice and uses that to create a set of sparse matricies. An example of a simulated ADF dataset with added noise and then denoised using a HOSVD decomposition. 

![Screen Shot 2020-05-07 at 8 30 33 AM](https://user-images.githubusercontent.com/41125831/81303708-abee0c00-9041-11ea-862d-df2bbd1eb15e.png)

We are interested in adding this to hyperspy, and feel it opens up an interesting conversation about tensor based and Higher Order Decompositions.  As to how this fits into the MVA class, these tensor based decompositions don't follow exactly the decomposition method which argues for creating a new series of tensor based decomposition methods. 

Talking with  @dnjohnstone yesterday he said that @tjof2 has worked with these tensor based decomposition approaches in the past and we would be interested in any comments for either of them.

**Comment by francisco-dlp** [comment]
Adding support for tensor decomposition method sound really good. Could you provide more details on why tensor decomposition methods don't fit into our current framework? Also, there are a number of python packages out there that implement several tensor decomposition methods. Which one(s) do you have in mind?

**Comment by dnjohnstone** [comment]
It's been a while since I looked at this - back in the day we used a Tucker decomposition I think very briefly, and I think it was a @tjof2 special, but I can't really remember the details because we stopped doing it quite quickly for various reasons.

Having a look around I thought this looked quite good: https://github.com/tensorly/tensorly

What do you think @francisco-dlp in terms of packages that are out there?

**Comment by tjof2** [comment]
I think we used scikit-tensor but that's stalled.

**Comment by tjof2** [comment]
@francisco-dlp asks a good question, I'd be interested in the answer.

The other thing would be to show some marked improvement over the matrix methods. For example, in your image above, does PCA on the unfolded dataset not work as well? Can you quantify it?

Also @francisco-dlp, in the unfolding, can we use the `iterpath='serpentine'` functionality at all? I wonder if that might help things further without resorting to tensors.

**Comment by CY-Zhang** [comment]
> Adding support for tensor decomposition method sound really good. Could you provide more details on why tensor decomposition methods don't fit into our current framework? Also, there are a number of python packages out there that implement several tensor decomposition methods. Which one(s) do you have in mind?

I think it worth mention that we are not just trying to decompose a tensor, but using a tensor decomposition method to find a low-rank approximation as the denoised result of input noisy tensor, similar to what get_decomposition_model(k) does on matrices. 

And based on my understanding, one major difference between finding a low-rank tensor and its matrix counterpart is that a low-rank tensor cannot be obtained from truncated full decomposition. This means we need to first determine the ranks we want for the low-rank tensor, then perform the decomposition (instead of matrix PCA where we can first do a full decomposition then truncate the decomposition). So what we are proposing is to first determine the ranks for each dimension by doing scree tests on unfolded matrices generated from the tensor, then do the decomposition to find low-rank tensor with given ranks. And I think this would be the main reason why tensor decomposition won't fit into the current decomposition scheme.

As for existing python packages with tensor decomposition capacity, one of them is the TensorLy package as @dnjohnstone  mentioned. Scikit-tensor seems to have some decomposition functions but I have never tried them. The method we applied to solve the decomposition problem was higher-order orthogonal iteration (HOOI), which has been proposed for quite a while and is a pretty standard tool to do tensor decomposition. I think TensorLy (probably also scikit-tensor) has implemented exactly the same tool to do tensor decomposition, but it does not have the rank determination part that we are proposing.

Our collaborator also has a newly developed tool to solve the decomposition problem (Zhang, Anru, and Rungang Han. "Optimal sparse singular value decomposition for high-dimensional high-order data." Journal of the American Statistical Association (2019): 1-34.), but the new method is very computationally expensive and seems to do little on denoising EM data.

**Comment by CY-Zhang** [comment]
> @francisco-dlp asks a good question, I'd be interested in the answer.
> 
> The other thing would be to show some marked improvement over the matrix methods. For example, in your image above, does PCA on the unfolded dataset not work as well? Can you quantify it?
> 
> Also @francisco-dlp, in the unfolding, can we use the `iterpath='serpentine'` functionality at all? I wonder if that might help things further without resorting to tensors.

That is a very good point and we are planning to discuss it in the manuscript that we are preparing. We have compared tensor SVD method to three other different denoising methods includes matrix PCA on the unfolded 2D matrix ( the other methods are non-local PCA and block-matching and filtering 4D)  using simulated data with known ground truth and different noise level. The denoising quality was quantified in terms of peak signal-to-noise ratio (PSNR) and the plot of output PSNR vs input PSNR is attached below. From the plot, you can see that the matrix PCA denoised result has several dB lower PSNR compare to the tensor SVD denoised results.

![DenoiseMethods](https://user-images.githubusercontent.com/10210880/81585994-8bdc8680-937a-11ea-9d4f-7ed844944e4c.png)

One thing is that we are proposing and testing the tensor SVD method on atomic-resolution 4D STEM data, as we think the ground truth can be well described by a low-rank tensor due to the repeating lattice structure. It is possible that matrix PCA on the unfolded dataset could work as well or even better than tensor SVD on some dataset that does not show this level of repeating structure.

**Comment by tjof2** [comment]
To clarify my question to @francisco-dlp - when we unfold the dataset prior to decomposition, do we/can we follow the "serpentine" pattern of (B) below, rather than "flyback" (A). For example, when you flatten an image from 2D matrix -> 1D vector, can we stack the pixels in the pattern of (B)? cf. #2240 

![image](https://user-images.githubusercontent.com/6626247/81674984-f9a1b480-9445-11ea-9f81-16c18c96530f.png)

This is relevant because I want to know if this, (B), helps preserve local structure in the data better than following (A), and leads to a better decomposition - and this is then relevant not only to this proposal (especially if you propose doing scree tests to determine the rank first), but more generally to how signals are unfolded.

**Comment by francisco-dlp** [comment]
@tjof2, that's obviously possible, albeit not implemented. I may be wrong, but I don't think that it'll make any different to SVD.

@CY-Zhang, when I last reviewed the bibliography of tensor decomposition (~3 years ago) there seemed to be an on-going debate about its benefits over matrix decomposition methods. Comparisons like the one that you have shared with us above seem to strongly suggest that there are clear benefits. However, I bet that you are comparing SVD on an unfolded dataset to tensor decomposition. Tensor decomposition exploits redundancies in all dimensions, and, therefore, the result should be superior in favorable conditions. However, wouldn't we obtain a similar result using a chain of matrix decompositions instead? Taking as an example the atomic resolution spectrum-image that you mention above, we could perform SVD on the unfolded dataset and then SVD again on the loadings (maps), exploiting the spatial redundancy. Wouldn't that be similar to a tensor decomposition?

**Comment by CY-Zhang** [comment]
@francisco-dlp Yes, my comparison was between tensor decomposition on 3D tensor and matrix SVD on a single matrix unfolded from the 3D tensor. I think the decomposition chain method you said is a good idea to denoise the 3D data and exploit redundancies along all three dimensions, and it might be beneficial to use different denoise ranks on different loading entrees, although this could be a little hard to implement. I will do some tests and see how it works.
